src/ShaderGen.Tests/ShaderBuiltinsTests.cs

```diff
@@ -0,0 +1,476 @@
+        {
+            // Creat failures list
+            List<Tuple<string, object, object>> failures = new List<Tuple<string, object, object>>();
+
+            // Get dictionary of fields by field name and type
+            Dictionary<Type, IReadOnlyCollection<FieldInfo>> childFieldInfos =
+                new Dictionary<Type, IReadOnlyCollection<FieldInfo>>();
+
+            Type currentType = typeof(T);
+            object aValue = a;
+            object bValue = b;
+            Stack<Tuple<string, Type, object, object>> stack = new Stack<Tuple<string, Type, object, object>>();
+            stack.Push(Tuple.Create(string.Empty, currentType, aValue, bValue));
+
+            while (stack.Count > 0)
+            {
+                // Pop top of stack.
+                Tuple<string, Type, object, object> tuple = stack.Pop();
+                currentType = tuple.Item2;
+                aValue = tuple.Item3;
+                bValue = tuple.Item4;
+
+                if (Equals(aValue, bValue)) continue;
+
+                // Get fields (cached)
+                if (!childFieldInfos.TryGetValue(currentType, out IReadOnlyCollection<FieldInfo> childFields))
+                    childFieldInfos.Add(currentType, childFields = currentType.GetFields().Where(f => !f.IsStatic).ToArray());
+
+                if (childFields.Count < 1)
+                {
+                    // No child fields, we have an inequality
+                    string fullName = tuple.Item1;
+                    failures.Add(Tuple.Create(fullName, aValue, bValue));
+                    continue;
+                }
+
+                foreach (FieldInfo childField in childFields)
+                {
+                    object aMemberValue = childField.GetValue(aValue);
+                    object bMemberValue = childField.GetValue(bValue);
+
+                    // Short cut equality
+                    if (Equals(aMemberValue, bMemberValue)) continue;
+
+                    string fullName = string.IsNullOrWhiteSpace(tuple.Item1)
+                        ? childField.Name
+                        : $"{tuple.Item1}.{childField.Name}";
+                    stack.Push(Tuple.Create(fullName, childField.FieldType, aMemberValue, bMemberValue));
+                }
+            }
+
+            return failures.AsReadOnly();
+        }
+
+
+        /// <summary>
+        /// The random number generators for each thread.
+        /// </summary>
+        private static readonly ThreadLocal<Random> _randomGenerators =
+            new ThreadLocal<Random>(() => new Random());
+
+        /// <summary>
+        /// Create a type with random data.
+        /// </summary>
+        /// <typeparam name="T">The random type</typeparam>
+        /// <param name="size">The optional number of bytes to fill.</param>
+        /// <returns></returns>
+        public unsafe T GetRandom<T>(int size = default(int))
```

Comment:
Since this only really works for float types, something like this would be much simpler:

```C#
public static unsafe T GenFloats<T>() where T : struct
{
    int floatCount = Unsafe.SizeOf<T>() / sizeof(float);
    float* floats = stackalloc[floatCount];
    for (int i = 0; i < floatCount; i++)
    {
        floats[i] = (float)random.NextDouble();
    }
    return Unsafe.Read<T>(floats);
}
```

Reply:
Stackalloc requires C# 7.3, do you want me to upgrade the project?

Reply:
It doesn't require 7.3 -- you might be thinking of stackalloc'd Spans.

Reply:
When I added the code into the test project I got a prompt asking me to upgrade it to 7.3.  I'll check again when I get back in a few hours.

Reply:
Thanks, I've replaced.